### ISSUE TITLE
paddle中的语言模型怎么在训练的时候计算整个句子的似然概率？

### ISSUE THREAD

**Issue by utopiar** [PaddlePaddle/Paddle#5541]
看generate_sequence_by_rnn_lm这个demo里是这么写的，最终infer的时候是得到最后一个词预测下一个词的概率，如下所示：
```python
    # fc(full connected) and output layer
    output = paddle.layer.fc(
        input=[rnn_cell], size=vocab_dim, act=paddle.activation.Softmax())

    if is_infer:
        last_word = paddle.layer.last_seq(input=output)
        return last_word
```
现在我想利用训练好的模型得到整个句子的似然概率，paddle里有没有对应的layer？

**Comment by NHZlX** [comment]
这个情感分析的例子可以看一下  http://book.paddlepaddle.org/index.cn.html

**Comment by utopiar** [comment]
是我没表达清楚吧，这个情感分析不是一个问题，我是想利用语言模型得到整个句子的似然概率，而不是一个分类概率！就是说我想利用语言模型计算文本的语言学特征，然后作为特征向上传

**Comment by lcy-seso** [comment]
既然可以得到下一个词的概率，直接自己在python脚本里面 score += log(p) 就可以了。不需要专门的layer来做。

**Comment by lcy-seso** [comment]
直接调用infer接口，走测试流程，不要用beam search，这个是文本生成用的，你需要的只是标准的预测流程，得到softmax的输出。自己把目标词所在类的概率取出来， score += log(p)。

**Comment by utopiar** [comment]
预测的时候确实按照 @lcy-seso 你的说法是可以得到的。
但我不单单是想在预测的时候得到，而是想将语言模型作为我整个model的一部分，利用它计算得到整个句子的似然概率然后传到上层网络，后面还想根据误差来回传调节我的语言模型，所以需要有专门的layer来计算`score += log(p)`

**Comment by lcy-seso** [comment]
那请在提问的时候，可能的描述你的需要。。。是有这样的cost的，但是只有你最后一个回复，才有一点知道你需要什么。。。

cross entropy 之后接 http://doc.paddlepaddle.org/doc_cn/api/v2/config/layer.html#sum-cost 。

**Comment by utopiar** [comment]
哈哈，谢谢~ 不好意思

**Comment by utopiar** [comment]
@lcy-seso  按照你的回答corss entropy直接接了sum_cost并没有对交叉熵得到的cost进行累加，后来测试发现需要吧交叉熵得到的cost利用trans转置后才能累加，这样在训练模型的时候batchsize是1的时候可以，如果bachsize是多个的时候就会把所有sample的cost累加能一个，请问怎么解决这个问题呢？

**Comment by lcy-seso** [comment]
在训练时候计算整个句子的概率是为了什么？优化还是评估？

**Comment by shibing624** [comment]
用于回传调节语言模型，具体的是用于计算cost。

**Comment by jacquesqiao** [comment]
标准的语言模型确实是预测下一个词的概率来计算的，请问你这个新的配置方法是一种新的模型么？有没有相关资料我们看下具体原理~

**Comment by lcy-seso** [comment]
语言模型训练时并不需要显示地计算困惑度，困惑度多用于评估。使用交叉熵完全等价与优化句子级别的困惑度（n-gram做了条件独立假设，lstm 隐含了句子级别的信息），参考models下面的例子（n-gram或者rnn language可以按需要选择）。

**Comment by utopiar** [comment]
是我们在调研的一种新的model，想结合语言模型和dssm框架去优化预训练的语言模型，利用dssm结合有标注的数据去调优语言模型，而在dssm里用rand_cost的时候需要有left-score和right-score，我们这里希望用句子的困惑度来表示score，所以需要在训练过程中计算困惑度

**Comment by lcy-seso** [comment]
sequence softmax + log + average pooling.

**Comment by shibing624** [comment]
可以详细解释一下 log + pooling 怎么做吗？还有sum pooling 是不是也可以？
